// Copyright 2020-2022 CYBERCRYPT

syntax = "proto3";

package d1.generic;
option go_package = "github.com/cybercryptio/d1-service-generic/v2/protobuf/generic";
option csharp_namespace = "CyberCrypt.D1.Client.Protobuf.Generic";
option java_package = "io.cybercrypt.d1.protobuf.generic";
option java_outer_classname = "GenericProto";

// Service for encryption and decryption of data.
service Generic {
  // Encrypts data and returns the ciphertext without storing it.
  rpc Encrypt (EncryptRequest) returns (EncryptResponse) {}

  // Authorizes the user for access permissions and if accessible, returns the decrypted content.
  rpc Decrypt (DecryptRequest) returns (DecryptResponse) {}
}

// Represents a request to encrypt data.
message EncryptRequest {
  // Data to encrypt.
  bytes plaintext = 1;
  // Associated data.
  bytes associated_data = 2;
  // Optional additional groups to add to the access list.
  repeated string group_ids = 3;
}

// Represents a response to an encryption request.
message EncryptResponse {
  // Ciphertext of the provided plaintext.
  bytes ciphertext = 1;
  // Associated data.
  bytes associated_data = 2;
  // The object ID of the encrypted data.
  string object_id = 3;
}

// Represents a request to decrypt data.
message DecryptRequest {
  // Data to decrypt.
  bytes ciphertext = 1;
  // Associated data.
  bytes associated_data = 2;
  // The object ID of the data.
  string object_id = 3;
}

// Represents a response to a decryption request.
message DecryptResponse {
  // Decrypted data.
  bytes plaintext = 1;
  // Associated data.
  bytes associated_data = 2;
}
